File: apps/kimi-code/src/tui/components/messages/thinking.ts
/**
 * Renders thinking content in the transcript.
 * Supports live in-place updates while thinking streams, then finalizes
 * without replacing the component.
 * Supports expand/collapse via Ctrl+O (shared with tool output).
 */

import { Text, truncateToWidth, type Component, type TUI } from '@moonshot-ai/pi-tui';

import {
  BRAILLE_SPINNER_FRAMES,
  BRAILLE_SPINNER_INTERVAL_MS,
  MESSAGE_INDENT,
  THINKING_PREVIEW_LINES,
} from '#/tui/constant/rendering';
import { STATUS_BULLET } from '#/tui/constant/symbols';
import { currentTheme } from '#/tui/theme';
import { isRenderCacheEnabled } from '#/tui/utils/render-cache';

export type ThinkingRenderMode = 'live' | 'finalized';

export class ThinkingComponent implements Component {
  private text: string;
  private showMarker: boolean;
  private mode: ThinkingRenderMode;
  private expanded = false;
  private hidden = false;
  private readonly ui: TUI | undefined;
  private spinnerFrame = 0;
  private spinnerInterval: ReturnType<typeof setInterval> | undefined;
  // Hold a single Text instance so pi-tui's (text, width) → lines cache
  // actually survives across renders. Re-constructing per render destroys
  // the cache and forces full re-wrap on every frame, which dominates CPU
  // once the transcript accumulates many finalized thinking blocks.
  private readonly textComponent: Text;

  private renderCache: { width: number; lines: string[] } | undefined;

  constructor(
    text: string,
    showMarker: boolean = true,
    mode: ThinkingRenderMode = 'finalized',
    ui?: TUI,
  ) {
    this.text = text;
    this.showMarker = showMarker;
    this.mode = mode;
    this.ui = ui;
    this.textComponent = new Text(this.styled(text), 0, 0);
    if (mode === 'live') {
      this.startSpinner();
    }
  }

  private markRenderDirty(): void {
    this.renderCache = undefined;
  }

  invalidate(): void {
    this.markRenderDirty();
    this.textComponent.setText(this.styled(this.text));
  }

  setText(text: string): void {
    if (this.text === text) return;
    this.text = text;
    this.markRenderDirty();
    this.textComponent.setText(this.styled(text));
  }

  private styled(text: string): string {
    return currentTheme.italicFg('textDim', text);
  }

  finalize(): void {
    this.mode = 'finalized';
    this.markRenderDirty();
    this.stopSpinner();
  }

  dispose(): void {
    this.stopSpinner();
  }

  setExpanded(expanded: boolean): void {
    if (this.expanded === expanded) return;
    this.expanded = expanded;
    this.markRenderDirty();
  }

  setHidden(hidden: boolean): void {
    if (this.hidden === hidden) return;
    this.hidden = hidden;
    this.markRenderDirty();
  }

  render(width: number): string[] {
    if (this.hidden && this.mode === 'finalized') return [];
    if (
      isRenderCacheEnabled() &&
      this.renderCache !== undefined &&
      this.renderCache.width === width
    ) {
      return this.renderCache.lines;
    }

    const contentWidth = Math.max(1, width - MESSAGE_INDENT.length);
    const contentLines = this.text.length > 0 ? this.textComponent.render(contentWidth) : [''];

    let rendered: string[];
    if (this.mode === 'live') {
      const visibleLines =
        contentLines.length > THINKING_PREVIEW_LINES
          ? contentLines.slice(contentLines.length - THINKING_PREVIEW_LINES)
          : contentLines;
      const spinner = currentTheme.fg(
        'textDim',
        `${BRAILLE_SPINNER_FRAMES[this.spinnerFrame] ?? BRAILLE_SPINNER_FRAMES[0]} `,
      );
      rendered = [
        '',
        spinner + currentTheme.fg('textDim', 'thinking...'),
        ...visibleLines.map((line) => MESSAGE_INDENT + line),
      ];
    } else {
      const lines: string[] = [''];
      for (let i = 0; i < contentLines.length; i++) {
        const p = i === 0 && this.showMarker ? currentTheme.fg('textDim', STATUS_BULLET) : MESSAGE_INDENT;
        lines.push(p + contentLines[i]);
      }

      if (this.expanded || contentLines.length <= THINKING_PREVIEW_LINES) {
        rendered = lines;
      } else {
        // Leading blank + first PREVIEW_LINES content lines + hint line.
        const truncated = lines.slice(0, 1 + THINKING_PREVIEW_LINES);
        const remaining = contentLines.length - THINKING_PREVIEW_LINES;
        const hint = `... (${String(remaining)} more lines, ctrl+o to expand)`;
        const indentWidth = Math.min(MESSAGE_INDENT.length, Math.max(0, width));
        const hintWidth = Math.max(0, width - indentWidth);
        truncated.push(
          ' '.repeat(indentWidth) + currentTheme.dim(truncateToWidth(hint, hintWidth, '…')),
        );
        rendered = truncated;
      }
    }

    if (isRenderCacheEnabled()) {
      this.renderCache = { width, lines: rendered };
    }
    return rendered;
  }

  private startSpinner(): void {
    if (this.ui === undefined || this.spinnerInterval !== undefined) return;
    this.spinnerInterval = setInterval(() => {
      this.spinnerFrame = (this.spinnerFrame + 1) % BRAILLE_SPINNER_FRAMES.length;
      this.markRenderDirty();
      this.ui?.requestRender();
    }, BRAILLE_SPINNER_INTERVAL_MS);
  }

  private stopSpinner(): void {
    if (this.spinnerInterval === undefined) return;
    clearInterval(this.spinnerInterval);
    this.spinnerInterval = undefined;
  }
}

File: crates/dimi-tui/src/components/messages/thinking.rs
//! `ThinkingComponent` — renders thinking content in the transcript
//! (port of `apps/dimi/src/tui/components/messages/thinking.ts`).

use crate::component::Component;
use crate::components::messages::{
    BRAILLE_SPINNER_FRAMES, MESSAGE_INDENT, STATUS_BULLET, THINKING_PREVIEW_LINES,
};
use crate::components::text::Text;
use crate::theme::{ColorToken, current_theme};
use crate::wrap::truncate_to_width;

/// Thinking render mode.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ThinkingMode {
    Live,
    Finalized,
}

/// Thinking content block with live in-place updates, expand/collapse, and
/// the live braille spinner header.
pub struct ThinkingComponent {
    text: String,
    show_marker: bool,
    mode: ThinkingMode,
    expanded: bool,
    hidden: bool,
    /// Hold a single Text instance so the (text, width) → lines cache
    /// survives across renders (same as TS).
    text_component: Text,
    render_cache: Option<(usize, Vec<String>)>,
}

impl ThinkingComponent {
    pub fn new(text: &str, show_marker: bool, mode: ThinkingMode) -> Self {
        let styled = current_theme().italic_fg(ColorToken::TextDim, text);
        ThinkingComponent {
            text: text.to_owned(),
            show_marker,
            mode,
            expanded: false,
            hidden: false,
            text_component: Text::new(&styled, 0, 0),
            render_cache: None,
        }
    }

    pub fn set_text(&mut self, text: &str) {
        if self.text == text {
            return;
        }
        self.text = text.to_owned();
        self.render_cache = None;
        let styled = current_theme().italic_fg(ColorToken::TextDim, text);
        self.text_component.set_text(&styled);
    }

    pub fn set_expanded(&mut self, expanded: bool) {
        if self.expanded == expanded {
            return;
        }
        self.expanded = expanded;
        self.render_cache = None;
    }

    pub fn set_hidden(&mut self, hidden: bool) {
        if self.hidden == hidden {
            return;
        }
        self.hidden = hidden;
        self.render_cache = None;
    }

    pub fn finalize(&mut self) {
        if self.mode == ThinkingMode::Finalized {
            return;
        }
        self.mode = ThinkingMode::Finalized;
        self.render_cache = None;
    }

    pub fn mode(&self) -> ThinkingMode {
        self.mode
    }
}

impl Component for ThinkingComponent {
    fn render(&mut self, width: usize) -> Vec<String> {
        if self.hidden && self.mode == ThinkingMode::Finalized {
            return Vec::new();
        }
        if let Some((cached_width, cached_lines)) = &self.render_cache {
            if *cached_width == width {
                return cached_lines.clone();
            }
        }

        let content_width = (width.saturating_sub(MESSAGE_INDENT.len())).max(1);
        let content_lines = if self.text.is_empty() {
            vec![String::new()]
        } else {
            self.text_component.render(content_width)
        };

        let rendered: Vec<String> = if self.mode == ThinkingMode::Live {
            let visible_lines = if content_lines.len() > THINKING_PREVIEW_LINES {
                content_lines[content_lines.len() - THINKING_PREVIEW_LINES..].to_vec()
            } else {
                content_lines
            };
            let spinner = current_theme().fg(
                ColorToken::TextDim,
                &format!("{} ", BRAILLE_SPINNER_FRAMES[0]),
            );
            let mut out = vec![String::new()];
            out.push(format!(
                "{spinner}{}",
                current_theme().fg(ColorToken::TextDim, "thinking...")
            ));
            for line in visible_lines {
                out.push(format!("{MESSAGE_INDENT}{line}"));
            }
            out
        } else {
            let mut lines: Vec<String> = vec![String::new()];
            for (i, content_line) in content_lines.iter().enumerate() {
                let p = if i == 0 && self.show_marker {
                    current_theme().fg(ColorToken::TextDim, STATUS_BULLET)
                } else {
                    MESSAGE_INDENT.to_owned()
                };
                lines.push(format!("{p}{content_line}"));
            }
            if self.expanded || content_lines.len() <= THINKING_PREVIEW_LINES {
                lines
            } else {
                // Leading blank + first PREVIEW_LINES content lines + hint line.
                let mut truncated = lines[..1 + THINKING_PREVIEW_LINES].to_vec();
                let remaining = content_lines.len() - THINKING_PREVIEW_LINES;
                let hint = format!("... ({remaining} more lines, ctrl+o to expand)");
                let indent_width = MESSAGE_INDENT.len().min(width);
                let hint_width = width.saturating_sub(indent_width);
                truncated.push(format!(
                    "{}{}",
                    " ".repeat(indent_width),
                    current_theme().dim(&truncate_to_width(&hint, hint_width, "…", false))
                ));
                truncated
            }
        };

        self.render_cache = Some((width, rendered.clone()));
        rendered
    }

    fn invalidate(&mut self) {
        self.render_cache = None;
        let styled = current_theme().italic_fg(ColorToken::TextDim, &self.text);
        self.text_component.set_text(&styled);
    }
}
